File: web/src/gf_apps/deprecated/__gf_publisher/ts/gf_posts_core/gf_utils.ts
/*
GloFlow application and media management/publishing platform
Copyright (C) 2024 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

///<reference path="../../../../d/jquery.d.ts" />

declare var gf_tagger__init_ui_v2;
declare var gf_tagger__http_add_tags_to_obj;

//---------------------------------------------------
// TAGGING_UI

export function init_tagging(p_post_id_str :string,
	p_post_element :HTMLDivElement,
	p_http_api_map,
	p_log_fun) {

	/*
	const http_api_map = {

		// GF_TAGGER
		"gf_tagger": {
			"add_tags_to_obj": async (p_new_tags_lst,
				p_obj_id_str,
				p_obj_type_str,
				p_tags_meta_map,
				p_log_fun)=>{
				const p = new Promise(async function(p_resolve_fun, p_reject_fun) {

					await gf_tagger__http_add_tags_to_obj(p_new_tags_lst,
						p_obj_id_str,
						p_obj_type_str,
						{}, // meta_map
						p_gf_host_str,
						p_log_fun);

					p_resolve_fun({
						"added_tags_lst": p_new_tags_lst,
					});
				});
				return p;
			}
		}
	};
	*/
	
	const obj_type_str = "post";

	const callbacks_map = {

		//---------------------------------------------------
		// TAGS
		//---------------------------------------------------
		"tags_pre_create_fun": async (p_tags_lst)=>{
			const p = new Promise(async function(p_resolve_fun, p_reject_fun) {

				// passing the image_id to the gf_tagger control via this callback allows for
				// customization of the image_id fetching mechanism (whether its in the template,
				// or fetched via rest api, etc., or pulled from some internal browser/web DB).
				p_resolve_fun(p_post_id_str);
			});
			return p;
		},
		
		//---------------------------------------------------
		"tags_created_fun": (p_tags_lst)=>{

			console.log("added tags >>>>>>>>>>>", p_tags_lst);

			p_tags_lst.forEach(p_tag_str=>{

				tag_display(p_tag_str);
			});
		},

		//---------------------------------------------------
		// NOTES
		//---------------------------------------------------
		"notes_pre_create_fun": (p_notes_lst)=>{
			const p = new Promise(async function(p_resolve_fun, p_reject_fun) {

				// passing the image_id to the gf_tagger control via this callback allows for
				// customization of the image_id fetching mechanism (whether its in the template,
				// or fetched via rest api, etc., or pulled from some internal browser/web DB).
				p_resolve_fun(p_post_id_str);
			});
			return p;
		},

		//---------------------------------------------------
		"notes_created_fun": (p_notes_lst)=>{

			console.log("added notes >>>>>>>>>>>", p_notes_lst)
		}

		//---------------------------------------------------
	}

	gf_tagger__init_ui_v2(p_post_id_str,
		obj_type_str,
		p_post_element,
		$("body"),
		callbacks_map,
		p_http_api_map,
		p_log_fun);

	//-------------------------------------------------
	function tag_display(p_tag_str) {

		/*
		check if the tags_container div exists, if not create it.
		the backend template has a div with class "tags_container" in the image container only if the image
		has tags. if it does not, the .tags_container div is not created
		*/
		if ($(p_post_element).find(".tags_container").length == 0) {
			$(p_post_element).append("<div class='tags_container'></div>");
		}

		$(p_post_element)
			.find(".tags_container")
			.append(`<a class='gf_image_tag' href='/v1/tags/objects?tag=${p_tag_str}&otype=image'>#${p_tag_str}</a>`)
	}

	//-------------------------------------------------
}

File: go/gf_identity/gf_identity_core/gf_utils.go
/*
GloFlow application and media management/publishing platform
Copyright (C) 2021 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

package gf_identity_core

import (
	// "fmt"
	"context"
	"net/http"
	"io/ioutil"
	"encoding/json"
	"github.com/gloflow/gloflow/go/gf_core"
)

//------------------------------------------------

type GFuserHTTPinputUpdate struct {
	ScreenNameStr       *string  `json:"screen_name_str" validate:"min=3,max=50"`
	EmailStr            *string  `json:"email_str"       validate:"min=6,max=50"`
	DescriptionStr      *string  `json:"description_str" validate:"min=1,max=2000"`

	ProfileImageURLstr *string `json:"profile_image_url_str" validate:"min=1,max=100"` // FIX!! - validation
	BannerImageURLstr  *string `json:"banner_image_url_str"  validate:"min=1,max=100"` // FIX!! - validation
}

type GFuserHTTPinputEmailConfirm struct {
	UserNameStr    GFuserName `validate:"required,min=3,max=50"`
	ConfirmCodeStr string     `validate:"required,min=10,max=20"`
}

type GFuserHTTPinputEmailLogin struct {
	EmailStr *string  `json:"email_str" validate:"min=6,max=50"`
}

type GFuserHTTPinputEmailLoginConfirm struct {
	UserNameStr    GFuserName `validate:"required,min=3,max=50"`
	ConfirmCodeStr string     `validate:"required,min=10,max=20"`
}

//---------------------------------------------------

func GetUserIDfromReq(pReq *http.Request,
	pAuthSubsystemTypeStr  string,
	pCtx		context.Context,
	pRuntimeSys *gf_core.RuntimeSys) (bool, gf_core.GF_ID, *gf_core.GFerror) {


	sessionID, sessionIDfoundBool := GetSessionID(pReq, pRuntimeSys)
	if !sessionIDfoundBool {
		return false, gf_core.GF_ID(""), nil
	}

	// AUTH0
	if pAuthSubsystemTypeStr == GF_AUTH_SUBSYSTEM_TYPE__AUTH0 {
		auth0session, gfErr := DBsqlAuth0getSession(sessionID, pCtx, pRuntimeSys)
		if gfErr != nil {
			return false, gf_core.GF_ID(""), gfErr
		}

		userID := auth0session.UserID

		return true, userID, nil
	}

	return false, gf_core.GF_ID(""), nil
}

//---------------------------------------------------

func ResolveUserName(pUserID gf_core.GF_ID,
	pCtx        context.Context,
	pRuntimeSys *gf_core.RuntimeSys) GFuserName {

	var userNameStr GFuserName

	/*
	LEGACY!! - old images dont have a user_id associated with them.
		before the user system was fully integrated into gf_images, images were added anonimously
		and did not have a user ID associated with them.
		for those images it is not possible to associate them with user_names. 
	*/
	if pUserID != "" {

		userID := pUserID

		resolvedUserNameStr, gfErr := DBsqlGetUserNameByID(userID, pCtx, pRuntimeSys)
		if gfErr != nil {

			// failing to resolve user_name should still return a user_name
			userNameStr = GFuserName("?")
			return userNameStr
		}
		userNameStr = resolvedUserNameStr
		
	} else {

		// IMPORTANT!! - pre-auth-system images are marked as owned by anonymous users.
		userNameStr = GFuserName("anon")
	}

	return userNameStr
}

//---------------------------------------------------
// CREATE_ID

func usersCreateID(pUserIdentifierStr string,
	pCreationUNIXtimeF float64) gf_core.GF_ID {

	fieldsForIDlst := []string{
		pUserIdentifierStr,
	}
	gfIDstr := gf_core.IDcreate(fieldsForIDlst,
		pCreationUNIXtimeF)

	return gfIDstr
}

//---------------------------------------------------
// GET_USER_NAME_FROM_CTX

func GetUserIDfromCtx(pCtx context.Context) (gf_core.GF_ID, bool) {
	
	userID := pCtx.Value("gf_user_id")

	if userID != nil {
		userID := gf_core.GF_ID(userID.(string))
		return userID, true
	}
	
	return gf_core.GF_ID(""), false
}

//---------------------------------------------------
// GET_SESSION_ID_FROM_CTX

func GetSessionIDfromCtx(pCtx context.Context) (gf_core.GF_ID, bool) {
	
	sessionIDval := pCtx.Value("gf_session_id")

	if sessionIDval != nil {
		sessionID := gf_core.GF_ID(sessionIDval.(string))
		return sessionID, true
	}
	
	return gf_core.GF_ID(""), false
}

//---------------------------------------------------
// HTTP
//---------------------------------------------------

func HTTPgetEmailLoginInput(pReq *http.Request) (*GFuserHTTPinputEmailLogin, *gf_core.GFerror) {
	input := &GFuserHTTPinputEmailLogin{}
	return input, nil
}

//---------------------------------------------------
// GET_AUTH_SUBSYSTEM_TYPE

func HTTPgetAuthSubsystemType(pReq *http.Request) string {
	authSubsystemTypeStr := pReq.Header.Get("gf_auth_type")
	return authSubsystemTypeStr
}

//---------------------------------------------------

func HTTPgetUserStdInput(pCtx context.Context,
	pReq        *http.Request,
	pRuntimeSys *gf_core.RuntimeSys) (map[string]interface{}, gf_core.GF_ID, GFuserAddressETH, *gf_core.GFerror) {

	inputMap, gfErr := gf_core.HTTPgetInput(pReq, pRuntimeSys)
	if gfErr != nil {
		return nil, "", GFuserAddressETH(""), gfErr
	}
	
	// user-name is supplied if the traditional auth system is used, and not web3/eth
	var userIDstr gf_core.GF_ID
	if inputUserIDstr, ok := inputMap["user_id_str"].(gf_core.GF_ID); ok {
		userIDstr = inputUserIDstr
	} else {

		// logged in users are added to context by gf_rpc, not supplied explicitly
		// via http request input (as they are for unauthenticated requests).
		userIDfromCtxStr, ok := GetUserIDfromCtx(pCtx) // pCtx.Value("gf_user_name").(string)
		if ok {
			userIDstr = userIDfromCtxStr
		}
	}

	pRuntimeSys.LogNewFun("DEBUG", "getting HTTP user std input", map[string]interface{}{"user_id_str": userIDstr,})

	// users eth address is used if the user picks that method instead of traditional
	var userAddressETHstr string;
	if inputUserAddressETHstr, ok := inputMap["user_address_eth_str"].(string); ok {
		userAddressETHstr = inputUserAddressETHstr
	}

	// one of the these values has to be supplied, they cant both be missing
	if userIDstr == "" && userAddressETHstr == "" {
		gfErr := gf_core.MongoHandleError("user_name_str or user_address_eth_str arguments are missing from request",
			"verify__input_data_missing_in_req_error",
			map[string]interface{}{},
			nil, "gf_identity_core", pRuntimeSys)
		return nil, "", GFuserAddressETH(""), gfErr
	}

	return inputMap, userIDstr, GFuserAddressETH(userAddressETHstr), nil
}

//---------------------------------------------------

func HTTPgetUserAddressETHinput(pReq *http.Request,
	pCtx        context.Context,
	pRuntimeSys *gf_core.RuntimeSys) (GFuserAddressETH, *gf_core.GFerror) {

	queryArgsMap := pReq.URL.Query()
	if valuesLst, ok := queryArgsMap["addr_eth"]; ok {
		return GFuserAddressETH(valuesLst[0]), nil
	} else {
		gfErr := gf_core.ErrorCreate("incoming http request is missing the addr_eth query-string arg",
			"verify__missing_key_error",
			map[string]interface{}{},
			nil, "gf_identity_core", pRuntimeSys)
		return GFuserAddressETH(""), gfErr
	}
	return GFuserAddressETH(""), nil
}

//---------------------------------------------------

func HTTPgetUserUpdateInput(pReq *http.Request,
	pRuntimeSys *gf_core.RuntimeSys) (*GFuserHTTPinputUpdate, *gf_core.GFerror) {

	handlerURLpathStr := pReq.URL.Path
	input             := GFuserHTTPinputUpdate{}
	bodyBytesLst, _   := ioutil.ReadAll(pReq.Body)
	err               := json.Unmarshal(bodyBytesLst, &input)
		
	if err != nil {
		gfErr := gf_core.ErrorCreate("failed to parse json http input for user update",
			"json_decode_error",
			map[string]interface{}{"handler_url_path_str": handlerURLpathStr,},
			err, "gf_identity_core", pRuntimeSys)
		return nil, gfErr
	}

	return &input, nil
}

//---------------------------------------------------
// GET_EMAIL_CONFIRM_INPUT

func HTTPgetEmailConfirmInput(pReq *http.Request,
	pRuntimeSys *gf_core.RuntimeSys) (*GFuserHTTPinputEmailConfirm, *gf_core.GFerror) {

	var userNameStr         string
	var confirmationCodeStr string

	queryArgsMap := pReq.URL.Query()
		
	//----------------------------
	// USER_NAME
	if valuesLst, ok := queryArgsMap["u"]; ok {
		userNameStr = valuesLst[0]
	} else {
		gfErr := gf_core.ErrorCreate("incoming http request is missing the email user_name query-string arg",
			"verify__missing_key_error",
			map[string]interface{}{},
			nil, "gf_identity_core", pRuntimeSys)
		return nil, gfErr
	}

	//----------------------------
	// CONFIRMATION_CODE
	if valuesLst, ok := queryArgsMap["c"]; ok {
		confirmationCodeStr = valuesLst[0]
	} else {
		gfErr := gf_core.ErrorCreate("incoming http request is missing the email confirmation_code query-string arg",
			"verify__missing_key_error",
			map[string]interface{}{},
			nil, "gf_identity_core", pRuntimeSys)
		return nil, gfErr
	}

	//----------------------------
	input := &GFuserHTTPinputEmailConfirm{
		UserNameStr:    GFuserName(userNameStr),
		ConfirmCodeStr: confirmationCodeStr,
	}

	return input, nil
}

//---------------------------------------------------
// GET_EMAIL_LOGIN_CONFIRM_INPUT

func HTTPgetEmailLoginConfirmInput(pReq *http.Request,
	pRuntimeSys *gf_core.RuntimeSys) (*GFuserHTTPinputEmailLoginConfirm, *gf_core.GFerror) {

	var userNameStr         string
	var confirmationCodeStr string

	queryArgsMap := pReq.URL.Query()
	
	if valuesLst, ok := queryArgsMap["u"]; ok {
		userNameStr = valuesLst[0]
	} else {
		gfErr := gf_core.ErrorCreate("incoming http request is missing the email user_name query-string arg",
			"verify__missing_key_error",
			map[string]interface{}{},
			nil, "gf_identity_core", pRuntimeSys)
		return nil, gfErr
	}

	if valuesLst, ok := queryArgsMap["c"]; ok {
		confirmationCodeStr = valuesLst[0]
	} else {
		gfErr := gf_core.ErrorCreate("incoming http request is missing the email confirmation_code query-string arg",
			"verify__missing_key_error",
			map[string]interface{}{},
			nil, "gf_identity_core", pRuntimeSys)
		return nil, gfErr
	}

	input := &GFuserHTTPinputEmailLoginConfirm{
		UserNameStr:    GFuserName(userNameStr),
		ConfirmCodeStr: confirmationCodeStr,
	}

	return input, nil
}

//---------------------------------------------------